lib.py

```diff
@@ -483,6 +518,84 @@ def convert_type(wdl_type: Any) -> Optional[WDLBase]:
             return None
         converted_inner_type = convert_type(inner_type)
         # if inner type conversion failed, then type is invalid, so return None
-        return outer_py_typ(converted_inner_type) if converted_inner_type is not None else None
+        return WDLArray(converted_inner_type, optional=optional, nonempty=nonempty) if converted_inner_type is not None else None
     # primitives remaining
-    return wdl_type_to_miniwdl_class(wdl_type)()
+    return wdl_type_to_miniwdl_class(wdl_type)(optional=optional)
+
+
+def test_gpu_available():
+    if bool(os.getenv("WDL_CONFORMANCE_TESTS__GPU")):
+        # override
+        return True
+    try:
+        p = subprocess.run("nvidia-smi".split(" "))
+    except (FileNotFoundError, subprocess.SubprocessError):
+        pass
+    else:
+        # we have an nvidia gpu
+        if p.returncode == 0:
+            return True
+    # This is copied from how Toil checks for amd gpus
+    # This may not work, I'm not sure what the current conventions for getting amd gpu data is
+    # see comment in src/toil/lib/accelerators.py::count_amd_gpus
+    try:
+        p = subprocess.run(["amd-smi", "static"])
+    except (FileNotFoundError, subprocess.SubprocessError):
+        pass
+    else:
+        if p.returncode == 0:
+            # we (maybe) have an amd gpu
+            return True
+    return False
+
+
+IGNORE_DEPENDENCIES = ["docker", "root", "singularity"]
+
+
+def test_dependencies(dependencies: Optional[List[str]], response: Dict[str, Any]) -> None:
+    """
+    Given a set of dependencies for a test, see if any of those dependencies are violated.
+    If so, change a failing test to a warning and update the reason.
```

Comment:
The test specification for the unit tests can be found at https://github.com/openwdl/wdl-tests/blob/main/docs/Specification.md. I linked the specification list instead of writing them in the file outright as it sounds like what is a "dependency" is flexible and not fully defined.

Response is just the runner's result after running the test, but I changed the semantics so it doesn't need the argument anymore.